NEW: Add support for Tip UI in TextField

Cheddam opened the pull request:
This PR adds schema and API to TextField allowing the Tip UI to be rendered alongside the field in React contexts, and goes hand-in-hand with the [admin PR implementing this UI in the InputField component](https://github.com/silverstripe/silverstripe-admin/pull/964).

See the Alt / Title fields in the [File details DSM document](https://projects.invisionapp.com/dsm/silver-stripe/silver-stripe/asset/components/5d70ab0cd81b4b445753d31b) for an example of the Tip UI in use.

This may make sense to expand to other fields over time, and it may also be smart to shift RightTitle into this UI - but for now, this is a standalone feature exclusive to basic TextFields.

maxime-rainville commented on src/Forms/TextField.php:
```diff
@@ -16,6 +16,26 @@ class TextField extends FormField
+    /**
+     * @var bool Whether the Tip UI should be rendered
+     */
+    protected $tipEnabled = false;
```
I think it would make more sense to encapsulate this logic in a Tip interface/class and pass that around.

This would allow us to use regular setters/getters to interact with the FormField tip. Also, if we decide we need to display tips in a different context latter on, we'll have a nice unified API to do so.

maxime-rainville commented on src/Forms/TextField.php:
```diff
@@ -16,6 +16,26 @@ class TextField extends FormField
+    /**
+     * @var string How important the tip is (normal or high). Informs the color and description.
+     */
+    protected $tipImportance = 'normal';
```
If we switch to an interface, this would be a good candidate for constants.

Cheddam commented:
@maxime-rainville I've extracted the Tip configuration into a separate object, and added some test coverage for it. Let me know what you think of the adjusted architecture!

maxime-rainville commented on src/Forms/TextField.php:
```diff
@@ -16,6 +16,11 @@ class TextField extends FormField
+    /**
+     * @var Tip|null A tip to render beside the input
+     */
+    protected $tip;
```
I think it would be fair to make this private and expect any sub-class to use or override the getter/setter if they need to do weird things.

maxime-rainville commented on src/Forms/TextField.php:
```diff
@@ -59,6 +64,29 @@ public function getMaxLength()
+    /**
+     * @return Tip|null
+     */
+    public function getTip()
```
```suggestion
    public function getTip(): Tip
```

maxime-rainville commented on src/Forms/TextField.php:
```diff
@@ -83,6 +111,11 @@ public function getSchemaDataDefaults()
         $data = parent::getSchemaDataDefaults();
         $data['data']['maxlength'] =  $this->getMaxLength();
+
+        if ($this->tip instanceof Tip) {
```
Generally, I prefer using the getter/setter rather than accessing private properties. That way there's never any doubt about  bypassing some validation rules or some magical accessor/mutator logic.

But that's just a personal preference.

ScopeyNZ replied:
Couple of ideas. 
- We can interface this (eg. `TippedFieldInterface`) that declares a `getTip` method? The `setTip` probably doesn't need to belong in the interface. Then you can just do `if ($this instanceof TippedFieldInterface` here.
- We can probably extract the implementation of `getTip` and `setTip` to a trait to save on code reuse. Then we can maybe move this logic to `FormField`?

ScopeyNZ replied:
The trait could also provide a helper method for adding a tip too - `->addTip(string $tip): Tip`?

Cheddam replied:
I've added a `TippableFieldInterface` which declares both `getTip(): ?Tip` and `setTip(?Tip $tip)`, as I'd rather see a consistent method of both getting and setting the tip, and `addTip` would suggest there could be multiple.

It still makes sense for this conditional to be checking against what `$this->getTip()` is returning, as it could be `null`, and if so we shouldn't try to use it. The interface might be helpful for actually determining whether this feature is available when a module developer tries to use it:

```php
if ($myTextField instanceof TippableFieldInterface) {
    $myTextField->setTip(...);
}
```

I think the trait will make sense if/when we spread this functionality to other fields, but I don't think it's necessary to create until then, as we can cleanly shift the methods / property out when it's appropriate (and it may not be useful at all if we end up shifting the logic up to `FormField`).

ScopeyNZ replied:
The only reason I recommended against having `setTip` in the interface as it seems like an unnecessary restriction on the implementor. What if I create a `InspriationalQuoteTippedField implements TippableFieldInterface` that creates a new `Tip` with an inspirational quote every time you call `getTip`?

ScopeyNZ replied:
I think it makes more sense for the interface to indicate how the object is used rather than how it might be configured - or at least those can be in different interfaces (but that's sounding a bit like interface overload at this point).

Cheddam replied:
😂 I appreciate the out-of-box thinking, but ultimately the goal of this feature is to allow someone to add their own extended description to a field. Having it in the interface means that the 'feature flag' example I provided will actually guarantee that the `setTip` call will work, even if it's ultimately a no-op for `InspirationalQuoteTippedField`.

That's also why I changed the name from `Tipped` to `Tippable`  - the former designates that it is always 'tipped', while the latter designates that it _can be_ 'tipped'.

maxime-rainville commented on src/Forms/TextField.php:
```diff
@@ -59,6 +64,29 @@ public function getMaxLength()
+     * @param Tip|null $tip The Tip to apply, or null to remove an existing one
+     * @return $this
+     */
+    public function setTip(Tip $tip = null)
```
```suggestion
    public function setTip(Tip $tip = null): self
```

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+ * Represents a Tip which can be rendered alongside a form field in the front-end.
+ * See the Tip component in the silverstripe/admin module.
+ *
+ * @package SilverStripe\Forms
```
```suggestion
```

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * These map to levels in the front-end Tip component
+     */
+    const IMPORTANCE_LEVELS = [
```
I would have made them two individual constants, but this way works as well.

Cheddam replied:
This more closely aligns to the front-end implementation (I agree that JS's objects are a bit more 'concrete' feeling than PHP's associative arrays though).

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+     */
+    private $message;
+
+    public function __construct(
```
```suggestion
   /**
     * @param string $message
     * @param string $importance_level
     * @param string $icon
     * @throws InvalidArgumentException
     */
    public function __construct(
```

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+     *
+     * @return array
+     */
+    public function getTipSchema()
```
```suggestion
    public function getTipSchema(): array
```

maxime-rainville replied:
Having a dedicated method for building the schema info is a good idea.

maxime-rainville replied:
Maybe consider using the getters when building the array.

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * @return string
+     */
+    public function getImportanceLevel()
```
```suggestion
    public function getImportanceLevel(): string
```

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * @param string $importance_level
+     */
+    public function setImportanceLevel($importance_level)
```
```suggestion
    public function setImportanceLevel(string $importance_level)
```

ScopeyNZ replied:
`void` or `return $this`?

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+     */
+    public function setImportanceLevel($importance_level)
+    {
+        if (!in_array($importance_level, self::IMPORTANCE_LEVELS)) {
```
Since you're using the exact same logic in the constructor, maybe just use the setter in the constructor.

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * @param string $icon
+     */
+    public function setIcon($icon)
```
```suggestion
    public function setIcon(string $icon)
```

ScopeyNZ replied:
Can we also either `return $this` or declare a `void` return type

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * @return string
+     */
+    public function getMessage()
```
```suggestion
    public function getMessage(): string
```

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * @param string $message
+     */
+    public function setMessage($message)
```
```suggestion
    public function setMessage(string $message)
```

ScopeyNZ replied:
`: void` or `return $this`?

maxime-rainville commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+
+    /**
+     * @param string $importance_level
+     */
```
```suggestion
     * @throws InvalidArgumentException
     */
```

ScopeyNZ commented on src/Forms/Tip.php:
```diff
@@ -0,0 +1,126 @@
+<?php
+
```
```suggestion
declare(strict_types=1);
```

ScopeyNZ commented:
Btw I think it's worth doing the typing stuff that Max and I have mentioned before merging. We might as well do typing fully now.

Cheddam commented:
I've applied types to all the affected methods, and added an interface to cover the functionality.
